hnn_core/network.py

```diff
@@ -459,14 +469,34 @@ def get_spike_rates(self):
         cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
         spike_rates = dict()
         all_spike_times = np.array(sum(self._times, []))
+        all_spike_types = np.array(sum(self._types, []))
+        all_spike_gids = np.array(sum(self._gids, []))
         tstart, tstop = min(all_spike_times), max(all_spike_times)
+
         for cell_type in cell_types:
-            trial_spike_rate = list()
-            for spike_times, spike_types in zip(self._times, self._types):
-                spike_times_cells = spike_times[spike_types == cell_type]
-                trial_spike_rate.append(
-                    len(spike_times_cells) / (tstop - tstart))
-            spike_rates[cell_type] = np.mean(trial_spike_rate)
+            type_mask = np.in1d(all_spike_types, cell_type)
+            cell_type_gids = np.unique(all_spike_gids[type_mask])
+            gid_spike_rate = np.zeros((len(self._times), len(cell_type_gids)))
+
+            trial_data = zip(self._types, self._gids)
+            for trial, (spike_types, spike_gids) in enumerate(trial_data):
+                trial_type_mask = np.in1d(spike_types, cell_type)
+                gid, gid_counts = np.unique(np.array(
+                    spike_gids)[trial_type_mask], return_counts=True)
+
+                gid_spike_rate[trial, cell_type_gids == gid] = (gid_counts / (
+                    tstop - tstart)) * 1000
```

Comment:
we still need to figure out where to get the `tstart` and `tstop` from. Should it be part of Spikes object or an argument to this method?

Reply:
As I think about it more it definitely makes sense for the `Spikes` to store this information. The interpretation of a spike train is very different if you can't differentiate inactivity from the recording being cut short.

Perhaps it can be an added to the instantiation as `self._bounds = (t_start, t_stop)`?

Reply:
I would just do `self.tstart` and `self.tstop` and then use that in the spike rate calculation. You'll also need to add tests to check they are set correctly when `Spikes` is an attribute of `Network`.